core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -420,4 +425,113 @@ protected WriteReplicaResult shardOperationOnReplica(BulkShardRequest request, I
         return next;
     }
 
+    public <Request extends ReplicatedWriteRequest<Request>, Response extends ReplicationResponse & WriteResponse>
+    WritePrimaryResult<Request, Response> executeSingleItemBulkRequestOnPrimary(
```

Comment:
wondering if we can put this something else .. maybe a static method on TransportIndexAction ? It's a shame to cluster this class with BWC code..